### PR TITLE
docs: Add article on configuring DPDK on VM

### PR DESCRIPTION
**Pull request by RamLavi**
This article is directly talking about D/S products, so it may move to
another location in the future.

### CODE REVIEW

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+Doing so will result in the creation of a [network-attachment-definition](https://docs.openshift.com/container-platform/4.12/virt/virtual_machines/vm_networking/virt-attaching-vm-multiple-networks.html) that will be used by the OpenShift Virtualization VM in the next steps.
+- [Checking](https://docs.openshift.com/container-platform/4.12/post_installation_configuration/machine-configuration-tasks.html#checking-mco-status_post-install-machine-configuration-tasks) the node has a machineConfigPool, and [creating](https://access.redhat.com/solutions/5688941) one if needed.
+- Creating a [performance-profile](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/cnf-create-performance-profiles.html) using the [Node-Tuning-Operator](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/using-node-tuning-operator.html).
+Doing so will create a [runtimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) that would be used by the OpenShift Virtualization VM in the next steps.
+
+In order to configure the cluster according to these steps please follow this OCP documentation.
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:35]: Missing a link that was present on the google doc:
```suggestion
In order to configure the cluster according to these steps please follow this [OCP documentation](https://docs.openshift.com/container-platform/4.12/networking/hardware_networks/using-dpdk-and-rdma.html).
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+## OpenShift Virtualization VM configuration
+
+### Kubevirt cluster wide configuration
+
+The pod running the VM (a.k.a. the virt-launcher pod) needs to run the DPDK optimized performance profile defined in the previous chapter. This is done by selecting the runtimeclassName in the pod running the VM.
+Since currently Kubevirt does not support setting the runtimeclassName per VM, the change needs to be on a cluster level, by using the [jsonpatch annotations](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/cluster-configuration.md#jsonpatch-annotations) on HCO:
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:42]: Could you please add a note regarding the effect of this setting?

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+...
+  domain:
+    interfaces:
+      - sriov: {}
+        name: nic-east
+        pciAddress: 0000:07:00.0
+  networks:
+    - name: nic-east
+      multus:
+        networkName: dpdk-net
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:73]: The indentation here is incorrect.

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+...
+```
+Note: In order to more easily identify and configure the SR-IOV interface inside the guest, NIC’s guest PCIAddress on the VM spec is set to 0000:07:00.0.
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:76]:
```suggestion
> *Note**
> In order to more easily identify and configure the SR-IOV interface inside the guest, NIC’s guest PCIAddress on the VM spec is set to `0000:07:00.0`.
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+#### CPU pinning
+
+To allow pinning a specific CPU quantity to the VM spec, the cpu snippet should look like this:
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:80]:
```suggestion
To allow pinning a specific CPU quantity to the VM spec, the cpu section should look like this:
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        cpu:
+          sockets: 1
+          threads: 2
+          cores: 5
+          dedicatedCpuPlacement: true
+...
+```
+Notes:
+- The sockets field should be 1, in order to make sure the CPUs are scheduled from the same NUMA Node.
+- In this snippet example, the VM will be scheduled with 5 [hyper-threads](https://access.redhat.com/articles/7445) (or 10 CPUs).
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:98]:
```
> *Note**
> 
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        resources:
+          requests:
+            memory: 8Gi
+        memory:
+          hugepages:
+            pageSize: "1Gi"
+...
+
+```
+Note: In this snippet we request 8 1Gi sized Hugepages.
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:142]:
```suggestion
Note: In this snippet we request eight 1Gi sized Hugepages.
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+kind: VirtualMachine
+metadata:
+  name: rhel9-dpdk-vm
+spec:
+  dataVolumeTemplates:
+    - apiVersion: cdi.kubevirt.io/v1beta1
+      kind: DataVolume
+      metadata:
+        creationTimestamp: null
+        name: rhel9-dpdk-vm
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:158]: Above it says we use RHEL 8

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+      spec:
+        sourceRef:
+          kind: DataSource
+          name: rhel9
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:162]: Same

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+      networks:
+        - name: default
+          pod: {}
+        - multus:
+            networkName: dpdk-net
+          name: nic-east
+      terminationGracePeriodSeconds: 180
+      volumes:
+        - dataVolume:
+            name: rhel9-dpdk-vm
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:214]: Same

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+## Guest Configuration
+
+Once the VM is deployed and running, there are configurations that need to be set once on the guest. The changes are explained on each subchapter with appropriate snippets. In the end, an example guest configuration summary script is offered.
+
+### CPU isolation on guest
+
+If we continue with the example from previous chapters, 5 hyper-threads translate to 10 CPUs on the guest:
+```bash
+$ cat /sys/fs/cgroup/cpuset/cpuset.cpus
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:233]: Could you please remove the `$` sign?
It is more convenient to copy the command using double-click when it is not there.

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+0-9
+
+```
+If you have housekeeping processes running on the guest then at least 1 hyper thread should be left as non-isolated. Assuming this is true, the first CPU and its sibling should not be isolated:
+```bash
+$ cat /sys/devices/system/cpu/cpu0/topology/core_cpus_list
```
Comment (orelmisan) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:239]: Same.

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
+# Configuring OpenShift Virtualization VM for DPDK workloads
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:1]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+## Introduction
+
+Running [DPDK](https://core.dpdk.org/doc/) workloads on OpenShift Virtualization VMs requires careful configuration steps, both on the cluster and guest levels. However, by following the guidelines provided in this article, users can achieve lower latency and higher throughput for packet processing in user space
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:5]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+## Prerequisites
+
+- [Deploy](https://www.redhat.com/en/technologies/cloud-computing/openshift/deploy-red-hat-openshift) an OpenShift Container Platform cluster (OCP v4.13)
+- [Deploy](https://docs.openshift.com/container-platform/4.12/virt/install/installing-virt-cli.html) OpenShift Virtualization plugin.tain
+- [Create](https://docs.openshift.com/container-platform/4.8/applications/projects/working-with-projects.html) a project on the cluster where the VM will run.
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:11]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 16]
```diff
@@ -0,0 +1,281 @@
+# Configuring OpenShift Virtualization VM for DPDK workloads
+
+## Introduction
+
+Running [DPDK](https://core.dpdk.org/doc/) workloads on OpenShift Virtualization VMs requires careful configuration steps, both on the cluster and guest levels. However, by following the guidelines provided in this article, users can achieve lower latency and higher throughput for packet processing in user space
+
+## Prerequisites
+
+- [Deploy](https://www.redhat.com/en/technologies/cloud-computing/openshift/deploy-red-hat-openshift) an OpenShift Container Platform cluster (OCP v4.13)
+- [Deploy](https://docs.openshift.com/container-platform/4.12/virt/install/installing-virt-cli.html) OpenShift Virtualization plugin.tain
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:10]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:11]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+- [Create](https://docs.openshift.com/container-platform/4.8/applications/projects/working-with-projects.html) a project on the cluster where the VM will run.
+- [Deploy](https://docs.openshift.com/container-platform/4.12/networking/hardware_networks/installing-sriov-operator.html) SR-IOV Network Operator.
+- The relevant worker nodes must have:
+  - SR-IOV enabled NICs.
+  - Allocatable Hugepages.
+
+*Note*: In this article an [Intel X710](https://www.dell.com/en-us/shop/intel-x710-quad-port-10gb-direct-attach-sfp-converged-network-adapter-full-height/apd/540-bbiw/networking#tabs_section) Network Adapter was used.
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:17]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+The configuration chapter will consist of three parts:
+- Cluster configuration - requires cluster-admin permissions.
+- VM spec configuration - requires project-admin permissions.
+- Guest configuration - based on RHEL8 OS.
+
+### Cluster configurations
+
+This step consists of:
+- Deploying a `vfio-pci` typed [SriovNetworkNodePolicy](https://docs.openshift.com/container-platform/4.12/networking/hardware_networks/configuring-sriov-device.html) and [SriovNetwork](https://docs.openshift.com/container-platform/4.12/networking/hardware_networks/configuring-sriov-device.html#cnf-assigning-a-sriov-network-to-a-vrf_configuring-sriov-device) using the [sriov-network-operator](https://github.com/openshift/sriov-network-operator) that is deployed on the cluster.
+Doing so will result in the creation of a [network-attachment-definition](https://docs.openshift.com/container-platform/4.12/virt/virtual_machines/vm_networking/virt-attaching-vm-multiple-networks.html) that will be used by the OpenShift Virtualization VM in the next steps.
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:30]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+- [Checking](https://docs.openshift.com/container-platform/4.12/post_installation_configuration/machine-configuration-tasks.html#checking-mco-status_post-install-machine-configuration-tasks) the node has a machineConfigPool, and [creating](https://access.redhat.com/solutions/5688941) one if needed.
+- Creating a [performance-profile](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/cnf-create-performance-profiles.html) using the [Node-Tuning-Operator](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/using-node-tuning-operator.html).
+Doing so will create a [runtimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) that would be used by the OpenShift Virtualization VM in the next steps.
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:33]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 1 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:35]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 3 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:73]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 5 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:80]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 6 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:98]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 7 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:142]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 8 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:158]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 9 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:162]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 10 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:214]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+        - cloudInitNoCloud:
+            userData: |-
+              #cloud-config
+              user: cloud-user
+              password: redhat
+              chpasswd: { expire: False }
+          name: cloudinitdisk
+```
+
+## Guest Configuration
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:225]: the checkup is running a testpmd application, requiring installing `dpdk` and `dpdk-tool`. But that's just a testing tool that is used to check DPDK workloads. 
In real life someone may use other tools or implement a DPDK application on their own. For example trex also uses testpmd but it self installs it.
Do you think I should add it anyway?

**Comment by RamLavi** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/a47762a736684cc8fbfe9a45798716e7c037d3ca..74a970ee2cd39ef5ab79253fc94f16168f48673c): Review fixes

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+#### CRI-O related annotations
+
+The following annotations are added, in order to ensure high performance on the [CRI-O](https://cri-o.io/) side. In order for these annotations to work, the VM underlying pod needs to run with the runtimeClassName configured on the previous chapters.
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    metadata:
+      Annotations:
```
Comment (phoracek) [inline docs/Configuring_openshift_virtualization_vm_for_dpdk.md:113]:
```suggestion
      annotations:
```

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:1]: The PR title still has the lower case "dpdk".

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 3 above)
Comment (orelmisan) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:76]: The indentation is still incorrect.

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 12 above)
Comment (orelmisan) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:247]: Could you please remove the `$` sign?
It is more convenient to copy the command using double-click when it is not there.

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:1]: changed the PR title as well

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 3 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:76]: I redid it now, but I don't think there was a problem. Are you sure there is an indentation issue?

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 21 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:113]: DONE

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 12 above)
Comment (RamLavi) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:247]: DONE (missed one)

File: docs/Configuring_openshift_virtualization_vm_for_dpdk.md (hunk 3 above)
Comment (orelmisan) [reply docs/Configuring_openshift_virtualization_vm_for_dpdk.md:76]: `domain` should be a child of `spec`:
http://kubevirt.io/api-reference/main/definitions.html#_v1_virtualmachineinstancespec

In this example it is his sibling becasue of incorrect indentation.

**Comment by orelmisan** [comment]
@RamLavi Several unrelated files were introduced on the last push.

**Review by orelmisan — APPROVED** [review]
Thank you for the changes @RamLavi

**Pull request merged**